{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%load_ext lab_black\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.append(\"..\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from funcs.utils import find_project_root"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Association"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "root = find_project_root()\n",
    "data_dir = root.parent / \"data\"\n",
    "assert data_dir.exists()\n",
    "ebi_dir = data_dir / \"gwas-catalog\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_path = ebi_dir / \"gwas_catalog_v1.0.2-studies_r2020-11-03.tsv\"\n",
    "assert df_path.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 10233 entries, 0 to 10232\n",
      "Data columns (total 16 columns):\n",
      " #   Column                      Non-Null Count  Dtype \n",
      "---  ------                      --------------  ----- \n",
      " 0   DATE ADDED TO CATALOG       10233 non-null  object\n",
      " 1   PUBMEDID                    10233 non-null  int64 \n",
      " 2   FIRST AUTHOR                10233 non-null  object\n",
      " 3   DATE                        10233 non-null  object\n",
      " 4   JOURNAL                     10233 non-null  object\n",
      " 5   LINK                        10233 non-null  object\n",
      " 6   STUDY                       10233 non-null  object\n",
      " 7   DISEASE/TRAIT               10233 non-null  object\n",
      " 8   INITIAL SAMPLE SIZE         10233 non-null  object\n",
      " 9   REPLICATION SAMPLE SIZE     3826 non-null   object\n",
      " 10  PLATFORM [SNPS PASSING QC]  10233 non-null  object\n",
      " 11  ASSOCIATION COUNT           10233 non-null  int64 \n",
      " 12  MAPPED_TRAIT                10221 non-null  object\n",
      " 13  MAPPED_TRAIT_URI            10221 non-null  object\n",
      " 14  STUDY ACCESSION             10233 non-null  object\n",
      " 15  GENOTYPING TECHNOLOGY       10233 non-null  object\n",
      "dtypes: int64(2), object(14)\n",
      "memory usage: 1.2+ MB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DATE ADDED TO CATALOG</th>\n",
       "      <th>PUBMEDID</th>\n",
       "      <th>FIRST AUTHOR</th>\n",
       "      <th>DATE</th>\n",
       "      <th>JOURNAL</th>\n",
       "      <th>LINK</th>\n",
       "      <th>STUDY</th>\n",
       "      <th>DISEASE/TRAIT</th>\n",
       "      <th>INITIAL SAMPLE SIZE</th>\n",
       "      <th>REPLICATION SAMPLE SIZE</th>\n",
       "      <th>PLATFORM [SNPS PASSING QC]</th>\n",
       "      <th>ASSOCIATION COUNT</th>\n",
       "      <th>MAPPED_TRAIT</th>\n",
       "      <th>MAPPED_TRAIT_URI</th>\n",
       "      <th>STUDY ACCESSION</th>\n",
       "      <th>GENOTYPING TECHNOLOGY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2012-10-19</td>\n",
       "      <td>21102463</td>\n",
       "      <td>Franke A</td>\n",
       "      <td>2010-11-21</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21102463</td>\n",
       "      <td>Genome-wide meta-analysis increases to 71 the ...</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>6,333 European ancestry cases, 15,056 European...</td>\n",
       "      <td>15,694 European ancestry cases, 14,026 Europea...</td>\n",
       "      <td>Affymetrix, Illumina [953241] (imputed)</td>\n",
       "      <td>71</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000384</td>\n",
       "      <td>GCST000879</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2011-01-13</td>\n",
       "      <td>21051598</td>\n",
       "      <td>Pereyra F</td>\n",
       "      <td>2010-11-04</td>\n",
       "      <td>Science</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/21051598</td>\n",
       "      <td>The major genetic determinants of HIV-1 contro...</td>\n",
       "      <td>HIV-1 control</td>\n",
       "      <td>516 European cases, 1,196 European controls, 3...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina [1384048] (imputed)</td>\n",
       "      <td>8</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000180</td>\n",
       "      <td>GCST000863</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2010-09-15</td>\n",
       "      <td>20700443</td>\n",
       "      <td>Meyer TE</td>\n",
       "      <td>2010-08-05</td>\n",
       "      <td>PLoS Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/20700443</td>\n",
       "      <td>Genome-wide association studies of serum magne...</td>\n",
       "      <td>Magnesium levels</td>\n",
       "      <td>15,366 European ancestry individuals</td>\n",
       "      <td>8,463 European ancestry individuals</td>\n",
       "      <td>Affymetrix, Illumina [~ 2500000] (imputed)</td>\n",
       "      <td>9</td>\n",
       "      <td>magnesium measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004845</td>\n",
       "      <td>GCST000756</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2009-01-12</td>\n",
       "      <td>19060906</td>\n",
       "      <td>Kathiresan S</td>\n",
       "      <td>2008-12-07</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/19060906</td>\n",
       "      <td>Common variants at 30 loci contribute to polyg...</td>\n",
       "      <td>HDL cholesterol</td>\n",
       "      <td>19,840 European ancestry individuals</td>\n",
       "      <td>Up to 20,623 European ancestry individuals</td>\n",
       "      <td>Affymetrix, Illumina [~ 2600000] (imputed)</td>\n",
       "      <td>14</td>\n",
       "      <td>high density lipoprotein cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004612</td>\n",
       "      <td>GCST000290</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2009-01-14</td>\n",
       "      <td>19060906</td>\n",
       "      <td>Kathiresan S</td>\n",
       "      <td>2008-12-07</td>\n",
       "      <td>Nat Genet</td>\n",
       "      <td>www.ncbi.nlm.nih.gov/pubmed/19060906</td>\n",
       "      <td>Common variants at 30 loci contribute to polyg...</td>\n",
       "      <td>LDL cholesterol</td>\n",
       "      <td>19,840 European ancestry individuals</td>\n",
       "      <td>Up to 20,623 European ancestry individuals</td>\n",
       "      <td>Affymetrix, Illumina [~ 2600000] (imputed)</td>\n",
       "      <td>11</td>\n",
       "      <td>low density lipoprotein cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004611</td>\n",
       "      <td>GCST000287</td>\n",
       "      <td>Genome-wide genotyping array</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  DATE ADDED TO CATALOG  PUBMEDID  FIRST AUTHOR        DATE     JOURNAL  \\\n",
       "0            2012-10-19  21102463      Franke A  2010-11-21   Nat Genet   \n",
       "1            2011-01-13  21051598     Pereyra F  2010-11-04     Science   \n",
       "2            2010-09-15  20700443      Meyer TE  2010-08-05  PLoS Genet   \n",
       "3            2009-01-12  19060906  Kathiresan S  2008-12-07   Nat Genet   \n",
       "4            2009-01-14  19060906  Kathiresan S  2008-12-07   Nat Genet   \n",
       "\n",
       "                                   LINK  \\\n",
       "0  www.ncbi.nlm.nih.gov/pubmed/21102463   \n",
       "1  www.ncbi.nlm.nih.gov/pubmed/21051598   \n",
       "2  www.ncbi.nlm.nih.gov/pubmed/20700443   \n",
       "3  www.ncbi.nlm.nih.gov/pubmed/19060906   \n",
       "4  www.ncbi.nlm.nih.gov/pubmed/19060906   \n",
       "\n",
       "                                               STUDY     DISEASE/TRAIT  \\\n",
       "0  Genome-wide meta-analysis increases to 71 the ...   Crohn's disease   \n",
       "1  The major genetic determinants of HIV-1 contro...     HIV-1 control   \n",
       "2  Genome-wide association studies of serum magne...  Magnesium levels   \n",
       "3  Common variants at 30 loci contribute to polyg...   HDL cholesterol   \n",
       "4  Common variants at 30 loci contribute to polyg...   LDL cholesterol   \n",
       "\n",
       "                                 INITIAL SAMPLE SIZE  \\\n",
       "0  6,333 European ancestry cases, 15,056 European...   \n",
       "1  516 European cases, 1,196 European controls, 3...   \n",
       "2               15,366 European ancestry individuals   \n",
       "3               19,840 European ancestry individuals   \n",
       "4               19,840 European ancestry individuals   \n",
       "\n",
       "                             REPLICATION SAMPLE SIZE  \\\n",
       "0  15,694 European ancestry cases, 14,026 Europea...   \n",
       "1                                                NaN   \n",
       "2                8,463 European ancestry individuals   \n",
       "3         Up to 20,623 European ancestry individuals   \n",
       "4         Up to 20,623 European ancestry individuals   \n",
       "\n",
       "                   PLATFORM [SNPS PASSING QC]  ASSOCIATION COUNT  \\\n",
       "0     Affymetrix, Illumina [953241] (imputed)                 71   \n",
       "1                Illumina [1384048] (imputed)                  8   \n",
       "2  Affymetrix, Illumina [~ 2500000] (imputed)                  9   \n",
       "3  Affymetrix, Illumina [~ 2600000] (imputed)                 14   \n",
       "4  Affymetrix, Illumina [~ 2600000] (imputed)                 11   \n",
       "\n",
       "                                       MAPPED_TRAIT  \\\n",
       "0                                   Crohn's disease   \n",
       "1                                   HIV-1 infection   \n",
       "2                             magnesium measurement   \n",
       "3  high density lipoprotein cholesterol measurement   \n",
       "4   low density lipoprotein cholesterol measurement   \n",
       "\n",
       "                       MAPPED_TRAIT_URI STUDY ACCESSION  \\\n",
       "0  http://www.ebi.ac.uk/efo/EFO_0000384      GCST000879   \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0000180      GCST000863   \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0004845      GCST000756   \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0004612      GCST000290   \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0004611      GCST000287   \n",
       "\n",
       "          GENOTYPING TECHNOLOGY  \n",
       "0  Genome-wide genotyping array  \n",
       "1  Genome-wide genotyping array  \n",
       "2  Genome-wide genotyping array  \n",
       "3  Genome-wide genotyping array  \n",
       "4  Genome-wide genotyping array  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0 = pd.read_csv(df_path, sep=\"\\t\")\n",
    "print(df0.info())\n",
    "df0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 10221 entries, 0 to 10232\n",
      "Data columns (total 4 columns):\n",
      " #   Column      Non-Null Count  Dtype \n",
      "---  ------      --------------  ----- \n",
      " 0   trait_id    10221 non-null  object\n",
      " 1   trait_name  10221 non-null  object\n",
      " 2   efo_term    10221 non-null  object\n",
      " 3   efo_id      10221 non-null  object\n",
      "dtypes: object(4)\n",
      "memory usage: 399.3+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_id</th>\n",
       "      <th>trait_name</th>\n",
       "      <th>efo_term</th>\n",
       "      <th>efo_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0-21102463</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1-21051598</td>\n",
       "      <td>HIV-1 control</td>\n",
       "      <td>HIV-1 infection</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2-20700443</td>\n",
       "      <td>Magnesium levels</td>\n",
       "      <td>magnesium measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3-19060906</td>\n",
       "      <td>HDL cholesterol</td>\n",
       "      <td>high density lipoprotein cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4-19060906</td>\n",
       "      <td>LDL cholesterol</td>\n",
       "      <td>low density lipoprotein cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004611</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     trait_id        trait_name  \\\n",
       "0  0-21102463   Crohn's disease   \n",
       "1  1-21051598     HIV-1 control   \n",
       "2  2-20700443  Magnesium levels   \n",
       "3  3-19060906   HDL cholesterol   \n",
       "4  4-19060906   LDL cholesterol   \n",
       "\n",
       "                                           efo_term  \\\n",
       "0                                   Crohn's disease   \n",
       "1                                   HIV-1 infection   \n",
       "2                             magnesium measurement   \n",
       "3  high density lipoprotein cholesterol measurement   \n",
       "4   low density lipoprotein cholesterol measurement   \n",
       "\n",
       "                                 efo_id  \n",
       "0  http://www.ebi.ac.uk/efo/EFO_0000384  \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0000180  \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0004845  \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0004612  \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0004611  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = (\n",
    "    df0.assign(\n",
    "        trait_id=lambda df: [\n",
    "            str(idx) + \"-\" + str(pubmedid)\n",
    "            for idx, pubmedid in enumerate(df[\"PUBMEDID\"].tolist())\n",
    "        ]\n",
    "    )\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"DISEASE/TRAIT\": \"trait_name\",\n",
    "            \"MAPPED_TRAIT\": \"efo_term\",\n",
    "            \"MAPPED_TRAIT_URI\": \"efo_id\",\n",
    "        }\n",
    "    )[[\"trait_id\", \"trait_name\", \"efo_term\", \"efo_id\"]]\n",
    "    .dropna()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "print(df.info())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "trait_name\n",
       "&beta;2-Glycoprotein I (&beta;2-GPI) plasma levels            1\n",
       "1,5-anhydroglucitol levels                                    1\n",
       "17-hydroxyprogesterone (17-OHP) levels                        1\n",
       "19,155 disease classifications in UK Biobank (pleiotropy)     1\n",
       "3-hydroxy-1-methylpropylmercapturic acid levels in smokers    1\n",
       "                                                             ..\n",
       "vWF and FVIII levels                                          2\n",
       "vWF levels                                                    1\n",
       "vWF levels in ischaemic stroke and hyperhomocysteinaemia      1\n",
       "vascular endothelial growth factor D levels                   1\n",
       "visceral adipose tissue/total adipose tissue ratio            1\n",
       "Length: 6339, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo = df[[\"trait_name\", \"efo_term\"]].drop_duplicates().groupby([\"trait_name\"]).size()\n",
    "foo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "trait_name\n",
       "Acute lymphoblastic leukemia (B-cell precursor)    2\n",
       "Adolescent idiopathic scoliosis                    2\n",
       "Advanced age-related macular degeneration          2\n",
       "Advanced glycation end-product levels              2\n",
       "Age of smoking initiation                          2\n",
       "                                                  ..\n",
       "Waist-hip ratio                                    2\n",
       "Waist-to-hip ratio adjusted for body mass index    2\n",
       "White blood cell count                             2\n",
       "White matter hyperintensity burden                 2\n",
       "vWF and FVIII levels                               2\n",
       "Length: 250, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "foo[foo > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait_id</th>\n",
       "      <th>trait_name</th>\n",
       "      <th>efo_term</th>\n",
       "      <th>efo_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1667</th>\n",
       "      <td>1667-21810271</td>\n",
       "      <td>vWF and FVIII levels</td>\n",
       "      <td>coagulation factor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5436</th>\n",
       "      <td>5436-30586737</td>\n",
       "      <td>vWF and FVIII levels</td>\n",
       "      <td>factor VIII measurement, von Willebrand factor...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004630, http://w...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           trait_id            trait_name  \\\n",
       "1667  1667-21810271  vWF and FVIII levels   \n",
       "5436  5436-30586737  vWF and FVIII levels   \n",
       "\n",
       "                                               efo_term  \\\n",
       "1667                     coagulation factor measurement   \n",
       "5436  factor VIII measurement, von Willebrand factor...   \n",
       "\n",
       "                                                 efo_id  \n",
       "1667               http://www.ebi.ac.uk/efo/EFO_0004634  \n",
       "5436  http://www.ebi.ac.uk/efo/EFO_0004630, http://w...  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df[df[\"trait_name\"] == \"Waist-hip ratio\"]\n",
    "df[df[\"trait_name\"] == \"vWF and FVIII levels\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Try assessing the trait-term mapping file instead"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trait term mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "trait_df_path = ebi_dir / \"gwas_catalog_trait-mappings_r2021-06-08.tsv\"\n",
    "assert trait_df_path.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 18263 entries, 0 to 18262\n",
      "Data columns (total 5 columns):\n",
      " #   Column         Non-Null Count  Dtype \n",
      "---  ------         --------------  ----- \n",
      " 0   Disease trait  18263 non-null  object\n",
      " 1   EFO term       18263 non-null  object\n",
      " 2   EFO URI        18263 non-null  object\n",
      " 3   Parent term    18263 non-null  object\n",
      " 4   Parent URI     18263 non-null  object\n",
      "dtypes: object(5)\n",
      "memory usage: 713.5+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease trait</th>\n",
       "      <th>EFO term</th>\n",
       "      <th>EFO URI</th>\n",
       "      <th>Parent term</th>\n",
       "      <th>Parent URI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clubfoot</td>\n",
       "      <td>familial clubfoot with or without associated l...</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_199315</td>\n",
       "      <td>Other disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alcohol dependence or heroin dependence or met...</td>\n",
       "      <td>methamphetamine dependence</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004701</td>\n",
       "      <td>Neurological disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Methamphetamine dependence</td>\n",
       "      <td>methamphetamine dependence</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004701</td>\n",
       "      <td>Neurological disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hypothyroidism</td>\n",
       "      <td>hypothyroidism</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004705</td>\n",
       "      <td>Other disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Age at menopause</td>\n",
       "      <td>age at menopause</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004704</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18258</th>\n",
       "      <td>Moderate or severe prolonged lymphopenia in di...</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Immune system disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18259</th>\n",
       "      <td>Severe prolonged lymphopenia in dimethyl fumar...</td>\n",
       "      <td>relapsing-remitting multiple sclerosis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003929</td>\n",
       "      <td>Immune system disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18260</th>\n",
       "      <td>Surgical necrotising enterocolitis in extremel...</td>\n",
       "      <td>necrotizing enterocolitis</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003928</td>\n",
       "      <td>Digestive system disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18261</th>\n",
       "      <td>Cognitive performance</td>\n",
       "      <td>neuropsychological test</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003926</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18262</th>\n",
       "      <td>Cognitive test performance</td>\n",
       "      <td>neuropsychological test</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0003926</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18263 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Disease trait  \\\n",
       "0                                               Clubfoot   \n",
       "1      Alcohol dependence or heroin dependence or met...   \n",
       "2                             Methamphetamine dependence   \n",
       "3                                         Hypothyroidism   \n",
       "4                                       Age at menopause   \n",
       "...                                                  ...   \n",
       "18258  Moderate or severe prolonged lymphopenia in di...   \n",
       "18259  Severe prolonged lymphopenia in dimethyl fumar...   \n",
       "18260  Surgical necrotising enterocolitis in extremel...   \n",
       "18261                              Cognitive performance   \n",
       "18262                         Cognitive test performance   \n",
       "\n",
       "                                                EFO term  \\\n",
       "0      familial clubfoot with or without associated l...   \n",
       "1                             methamphetamine dependence   \n",
       "2                             methamphetamine dependence   \n",
       "3                                         hypothyroidism   \n",
       "4                                       age at menopause   \n",
       "...                                                  ...   \n",
       "18258             relapsing-remitting multiple sclerosis   \n",
       "18259             relapsing-remitting multiple sclerosis   \n",
       "18260                          necrotizing enterocolitis   \n",
       "18261                            neuropsychological test   \n",
       "18262                            neuropsychological test   \n",
       "\n",
       "                                         EFO URI                Parent term  \\\n",
       "0      http://www.orpha.net/ORDO/Orphanet_199315              Other disease   \n",
       "1           http://www.ebi.ac.uk/efo/EFO_0004701      Neurological disorder   \n",
       "2           http://www.ebi.ac.uk/efo/EFO_0004701      Neurological disorder   \n",
       "3           http://www.ebi.ac.uk/efo/EFO_0004705              Other disease   \n",
       "4           http://www.ebi.ac.uk/efo/EFO_0004704          Other measurement   \n",
       "...                                          ...                        ...   \n",
       "18258       http://www.ebi.ac.uk/efo/EFO_0003929     Immune system disorder   \n",
       "18259       http://www.ebi.ac.uk/efo/EFO_0003929     Immune system disorder   \n",
       "18260       http://www.ebi.ac.uk/efo/EFO_0003928  Digestive system disorder   \n",
       "18261       http://www.ebi.ac.uk/efo/EFO_0003926          Other measurement   \n",
       "18262       http://www.ebi.ac.uk/efo/EFO_0003926          Other measurement   \n",
       "\n",
       "                                 Parent URI  \n",
       "0      http://www.ebi.ac.uk/efo/EFO_0000408  \n",
       "1      http://www.ebi.ac.uk/efo/EFO_0000618  \n",
       "2      http://www.ebi.ac.uk/efo/EFO_0000618  \n",
       "3      http://www.ebi.ac.uk/efo/EFO_0000408  \n",
       "4      http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "...                                     ...  \n",
       "18258  http://www.ebi.ac.uk/efo/EFO_0000540  \n",
       "18259  http://www.ebi.ac.uk/efo/EFO_0000540  \n",
       "18260  http://www.ebi.ac.uk/efo/EFO_0000405  \n",
       "18261  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "18262  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "\n",
       "[18263 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0 = pd.read_csv(trait_df_path, sep=\"\\t\")\n",
    "print(df0.info())\n",
    "df0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Other measurement                   11677\n",
       "Neurological disorder                 912\n",
       "Lipid or lipoprotein measurement      705\n",
       "Response to drug                      664\n",
       "Other disease                         613\n",
       "Other trait                           574\n",
       "Cancer                                534\n",
       "Biological process                    412\n",
       "Cardiovascular disease                368\n",
       "Cardiovascular measurement            318\n",
       "Immune system disorder                309\n",
       "Metabolic disorder                    295\n",
       "Body measurement                      294\n",
       "Digestive system disorder             216\n",
       "Hematological measurement             211\n",
       "Inflammatory measurement              113\n",
       "Liver enzyme measurement               48\n",
       "Name: Parent term, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0[\"Parent term\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13408\n"
     ]
    }
   ],
   "source": [
    "print(df0[\"Disease trait\"].drop_duplicates().pipe(len))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1      10637\n",
       "2       2050\n",
       "3        494\n",
       "4        113\n",
       "5         40\n",
       "6         17\n",
       "7         12\n",
       "8         11\n",
       "11         9\n",
       "9          6\n",
       "10         4\n",
       "15         2\n",
       "16         2\n",
       "12         1\n",
       "235        1\n",
       "19         1\n",
       "13         1\n",
       "66         1\n",
       "18         1\n",
       "53         1\n",
       "329        1\n",
       "41         1\n",
       "22         1\n",
       "55         1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_mapping = (\n",
    "    df0[[\"Disease trait\", \"EFO term\"]].drop_duplicates().groupby(\"Disease trait\").size()\n",
    ")\n",
    "check_mapping.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Disease trait\n",
       "Blood protein levels    329\n",
       "dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_mapping[check_mapping == 329]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Disease trait\n",
       "Age-related disease endophenotypes                                 9\n",
       "Alzheimer's disease or cardiovascular risk factors (pleiotropy)    9\n",
       "Glucose homeostasis traits                                         9\n",
       "Red blood cell traits                                              9\n",
       "Serum metabolite levels                                            9\n",
       "Sex hormone levels                                                 9\n",
       "dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_mapping[check_mapping == 9]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease trait</th>\n",
       "      <th>EFO term</th>\n",
       "      <th>EFO URI</th>\n",
       "      <th>Parent term</th>\n",
       "      <th>Parent URI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>SPARC-like protein 1 measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008289</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>small nuclear ribonucleoprotein F measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008288</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>tenascin measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008296</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>T-lymphocyte surface antigen Ly-9 measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008295</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>thioredoxin domain-containing protein 12 measu...</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0008298</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17130</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>galectin-3-binding protein measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0010241</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17131</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>beta globulin measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0010242</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17468</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>soluble transferrin receptor measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004460</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17870</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>ferritin measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004459</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17877</th>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>c-reactive protein measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004458</td>\n",
       "      <td>Inflammatory measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004872</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>329 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Disease trait  \\\n",
       "108    Blood protein levels   \n",
       "109    Blood protein levels   \n",
       "110    Blood protein levels   \n",
       "111    Blood protein levels   \n",
       "112    Blood protein levels   \n",
       "...                     ...   \n",
       "17130  Blood protein levels   \n",
       "17131  Blood protein levels   \n",
       "17468  Blood protein levels   \n",
       "17870  Blood protein levels   \n",
       "17877  Blood protein levels   \n",
       "\n",
       "                                                EFO term  \\\n",
       "108                     SPARC-like protein 1 measurement   \n",
       "109        small nuclear ribonucleoprotein F measurement   \n",
       "110                                 tenascin measurement   \n",
       "111        T-lymphocyte surface antigen Ly-9 measurement   \n",
       "112    thioredoxin domain-containing protein 12 measu...   \n",
       "...                                                  ...   \n",
       "17130             galectin-3-binding protein measurement   \n",
       "17131                          beta globulin measurement   \n",
       "17468           soluble transferrin receptor measurement   \n",
       "17870                               ferritin measurement   \n",
       "17877                     c-reactive protein measurement   \n",
       "\n",
       "                                    EFO URI               Parent term  \\\n",
       "108    http://www.ebi.ac.uk/efo/EFO_0008289         Other measurement   \n",
       "109    http://www.ebi.ac.uk/efo/EFO_0008288         Other measurement   \n",
       "110    http://www.ebi.ac.uk/efo/EFO_0008296         Other measurement   \n",
       "111    http://www.ebi.ac.uk/efo/EFO_0008295         Other measurement   \n",
       "112    http://www.ebi.ac.uk/efo/EFO_0008298         Other measurement   \n",
       "...                                     ...                       ...   \n",
       "17130  http://www.ebi.ac.uk/efo/EFO_0010241         Other measurement   \n",
       "17131  http://www.ebi.ac.uk/efo/EFO_0010242         Other measurement   \n",
       "17468  http://www.ebi.ac.uk/efo/EFO_0004460         Other measurement   \n",
       "17870  http://www.ebi.ac.uk/efo/EFO_0004459         Other measurement   \n",
       "17877  http://www.ebi.ac.uk/efo/EFO_0004458  Inflammatory measurement   \n",
       "\n",
       "                                 Parent URI  \n",
       "108    http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "109    http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "110    http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "111    http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "112    http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "...                                     ...  \n",
       "17130  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "17131  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "17468  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "17870  http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "17877  http://www.ebi.ac.uk/efo/EFO_0004872  \n",
       "\n",
       "[329 rows x 5 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0[df0[\"Disease trait\"] == \"Blood protein levels\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease trait</th>\n",
       "      <th>EFO term</th>\n",
       "      <th>EFO URI</th>\n",
       "      <th>Parent term</th>\n",
       "      <th>Parent URI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2991</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>high density lipoprotein cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004612</td>\n",
       "      <td>Lipid or lipoprotein measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3652</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>systolic blood pressure</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0006335</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3709</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>diastolic blood pressure</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0006336</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6680</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>ventricular rate measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0007928</td>\n",
       "      <td>Cardiovascular measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9159</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>hematocrit</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004348</td>\n",
       "      <td>Hematological measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9206</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>body mass index</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004340</td>\n",
       "      <td>Body measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16430</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>triglyceride measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004530</td>\n",
       "      <td>Lipid or lipoprotein measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16688</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>total cholesterol measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004574</td>\n",
       "      <td>Lipid or lipoprotein measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17445</th>\n",
       "      <td>Age-related disease endophenotypes</td>\n",
       "      <td>glucose measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004468</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Disease trait  \\\n",
       "2991   Age-related disease endophenotypes   \n",
       "3652   Age-related disease endophenotypes   \n",
       "3709   Age-related disease endophenotypes   \n",
       "6680   Age-related disease endophenotypes   \n",
       "9159   Age-related disease endophenotypes   \n",
       "9206   Age-related disease endophenotypes   \n",
       "16430  Age-related disease endophenotypes   \n",
       "16688  Age-related disease endophenotypes   \n",
       "17445  Age-related disease endophenotypes   \n",
       "\n",
       "                                               EFO term  \\\n",
       "2991   high density lipoprotein cholesterol measurement   \n",
       "3652                            systolic blood pressure   \n",
       "3709                           diastolic blood pressure   \n",
       "6680                       ventricular rate measurement   \n",
       "9159                                         hematocrit   \n",
       "9206                                    body mass index   \n",
       "16430                          triglyceride measurement   \n",
       "16688                     total cholesterol measurement   \n",
       "17445                               glucose measurement   \n",
       "\n",
       "                                    EFO URI                       Parent term  \\\n",
       "2991   http://www.ebi.ac.uk/efo/EFO_0004612  Lipid or lipoprotein measurement   \n",
       "3652   http://www.ebi.ac.uk/efo/EFO_0006335                 Other measurement   \n",
       "3709   http://www.ebi.ac.uk/efo/EFO_0006336                 Other measurement   \n",
       "6680   http://www.ebi.ac.uk/efo/EFO_0007928        Cardiovascular measurement   \n",
       "9159   http://www.ebi.ac.uk/efo/EFO_0004348         Hematological measurement   \n",
       "9206   http://www.ebi.ac.uk/efo/EFO_0004340                  Body measurement   \n",
       "16430  http://www.ebi.ac.uk/efo/EFO_0004530  Lipid or lipoprotein measurement   \n",
       "16688  http://www.ebi.ac.uk/efo/EFO_0004574  Lipid or lipoprotein measurement   \n",
       "17445  http://www.ebi.ac.uk/efo/EFO_0004468                 Other measurement   \n",
       "\n",
       "                                 Parent URI  \n",
       "2991   http://www.ebi.ac.uk/efo/EFO_0004732  \n",
       "3652   http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "3709   http://www.ebi.ac.uk/efo/EFO_0001444  \n",
       "6680   http://www.ebi.ac.uk/efo/EFO_0004298  \n",
       "9159   http://www.ebi.ac.uk/efo/EFO_0004503  \n",
       "9206   http://www.ebi.ac.uk/efo/EFO_0004324  \n",
       "16430  http://www.ebi.ac.uk/efo/EFO_0004529  \n",
       "16688  http://www.ebi.ac.uk/efo/EFO_0004732  \n",
       "17445  http://www.ebi.ac.uk/efo/EFO_0001444  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0[df0[\"Disease trait\"] == \"Age-related disease endophenotypes\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- produce the following\n",
    "  - gwas-to-efo 'edge' list\n",
    "- cleaning steps: \n",
    "  - remove mappings where a trait maps to more than `5` efo terms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "SIZE_LIMIT_MULTI_MAPPING = 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 18263 entries, 0 to 18262\n",
      "Data columns (total 5 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   gwas_name        18263 non-null  object\n",
      " 1   efo_name         18263 non-null  object\n",
      " 2   efo_id           18263 non-null  object\n",
      " 3   parent_efo_name  18263 non-null  object\n",
      " 4   parent_efo_id    18263 non-null  object\n",
      "dtypes: object(5)\n",
      "memory usage: 713.5+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "df1 = df0.rename(\n",
    "    columns={\n",
    "        \"Disease trait\": \"gwas_name\",\n",
    "        \"EFO term\": \"efo_name\",\n",
    "        \"EFO URI\": \"efo_id\",\n",
    "        \"Parent term\": \"parent_efo_name\",\n",
    "        \"Parent URI\": \"parent_efo_id\",\n",
    "    }\n",
    ")\n",
    "print(df1.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "74\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Disease trait\n",
       "Age-related disease endophenotypes                                          9\n",
       "Age-related diseases and mortality                                          7\n",
       "Age-related diseases, mortality and associated endophenotypes              16\n",
       "Alzheimer's disease or cardiovascular risk factors (pleiotropy)             9\n",
       "Amyotrophic lateral sclerosis and immune-mediated diseases (pleiotropy)     7\n",
       "dtype: int64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exclude_traits = check_mapping[check_mapping > SIZE_LIMIT_MULTI_MAPPING]\n",
    "print(len(exclude_traits))\n",
    "exclude_traits[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Age-related disease endophenotypes',\n",
       " 'Age-related diseases and mortality',\n",
       " 'Age-related diseases, mortality and associated endophenotypes',\n",
       " \"Alzheimer's disease or cardiovascular risk factors (pleiotropy)\",\n",
       " 'Amyotrophic lateral sclerosis and immune-mediated diseases (pleiotropy)']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exclude_traits_list = list(exclude_traits.index)\n",
    "exclude_traits_list[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Int64Index: 16871 entries, 0 to 18262\n",
      "Data columns (total 5 columns):\n",
      " #   Column           Non-Null Count  Dtype \n",
      "---  ------           --------------  ----- \n",
      " 0   gwas_name        16871 non-null  object\n",
      " 1   efo_name         16871 non-null  object\n",
      " 2   efo_id           16871 non-null  object\n",
      " 3   parent_efo_name  16871 non-null  object\n",
      " 4   parent_efo_id    16871 non-null  object\n",
      "dtypes: object(5)\n",
      "memory usage: 790.8+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gwas_name</th>\n",
       "      <th>efo_name</th>\n",
       "      <th>efo_id</th>\n",
       "      <th>parent_efo_name</th>\n",
       "      <th>parent_efo_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clubfoot</td>\n",
       "      <td>familial clubfoot with or without associated l...</td>\n",
       "      <td>http://www.orpha.net/ORDO/Orphanet_199315</td>\n",
       "      <td>Other disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alcohol dependence or heroin dependence or met...</td>\n",
       "      <td>methamphetamine dependence</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004701</td>\n",
       "      <td>Neurological disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Methamphetamine dependence</td>\n",
       "      <td>methamphetamine dependence</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004701</td>\n",
       "      <td>Neurological disorder</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Hypothyroidism</td>\n",
       "      <td>hypothyroidism</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004705</td>\n",
       "      <td>Other disease</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0000408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Age at menopause</td>\n",
       "      <td>age at menopause</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0004704</td>\n",
       "      <td>Other measurement</td>\n",
       "      <td>http://www.ebi.ac.uk/efo/EFO_0001444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           gwas_name  \\\n",
       "0                                           Clubfoot   \n",
       "1  Alcohol dependence or heroin dependence or met...   \n",
       "2                         Methamphetamine dependence   \n",
       "3                                     Hypothyroidism   \n",
       "4                                   Age at menopause   \n",
       "\n",
       "                                            efo_name  \\\n",
       "0  familial clubfoot with or without associated l...   \n",
       "1                         methamphetamine dependence   \n",
       "2                         methamphetamine dependence   \n",
       "3                                     hypothyroidism   \n",
       "4                                   age at menopause   \n",
       "\n",
       "                                      efo_id        parent_efo_name  \\\n",
       "0  http://www.orpha.net/ORDO/Orphanet_199315          Other disease   \n",
       "1       http://www.ebi.ac.uk/efo/EFO_0004701  Neurological disorder   \n",
       "2       http://www.ebi.ac.uk/efo/EFO_0004701  Neurological disorder   \n",
       "3       http://www.ebi.ac.uk/efo/EFO_0004705          Other disease   \n",
       "4       http://www.ebi.ac.uk/efo/EFO_0004704      Other measurement   \n",
       "\n",
       "                          parent_efo_id  \n",
       "0  http://www.ebi.ac.uk/efo/EFO_0000408  \n",
       "1  http://www.ebi.ac.uk/efo/EFO_0000618  \n",
       "2  http://www.ebi.ac.uk/efo/EFO_0000618  \n",
       "3  http://www.ebi.ac.uk/efo/EFO_0000408  \n",
       "4  http://www.ebi.ac.uk/efo/EFO_0001444  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_to_efo = df1[df1[\"gwas_name\"].apply(lambda x: x not in exclude_traits_list)]\n",
    "print(gwas_to_efo.info())\n",
    "gwas_to_efo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file = ebi_dir / \"clean_gwas_to_efo.csv\"\n",
    "gwas_to_efo.to_csv(output_file, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
